### PR TITLE
Use SOCK_NONBLOCK to reduce system calls for outgoing connections

### PR DESCRIPTION
**Pull request by panjf2000**
### What this PR mainly does is:
1. Refactor the `anetCreateSocket()` to make it more generic for more socket arguments, and use `SOCK_NONBLOCK` if available, which will reduce two system calls (`F_GETFL` and `F_SETFL`) of enabling the non-blocking mode on each newly created socket.
2. Clean up the unused `anetUnixGenericConnect()` that calls `anetCreateSocket()`.

`SOCK_NONBLOCK` for system call `socket()` is supported on most UNIX-like platforms (`linux`, `dragonfly`, `freebsd`, `netbsd`, `openbsd`, `solaris`, etc.). This improvement will significantly reduce the system calls considering how massively `anetTcpGenericConnect()` will be called when needed.

As for the cleanup, `anetUnixGenericConnect` was introduced in c61e692 and the only reference back then was from the `createClient()` in `redis-benchmark.c` which had been removed in ec8f066 and made it the dead code. Most of that dead code was also cleaned up in f657315e, and it seems that the `anetUnixGenericConnect` got left out. Therefore, I also cleaned it up, ***but I'm not so certain about doing this cleanup in this PR. Maybe you would prefer to do it in a separate PR?***

### References
- [socket(2) on Linux](https://man7.org/linux/man-pages/man2/socket.2.html)
- [socket(2) on FreeBSD](https://man.freebsd.org/cgi/man.cgi?socket(2))
- [socket(2) on DragonFly](https://man.dragonflybsd.org/?command=socket&section=2)
- [socket(2) on NetBSD](https://man.netbsd.org/socket.2)
- [socket(2) on OpenBSD](https://man.openbsd.org/socket.2)
- [socket(3c) on Solaris](https://docs.oracle.com/cd/E88353_01/html/E37843/socket-3c.html)
- [socket(3socket) on illumos](https://illumos.org/man/3SOCKET/socket)

### CODE REVIEW

**Comment by panjf2000** [comment]
@madolson @zuiderkwast Got a minute for this?

File: src/anet.c [hunk 1]
```diff
@@ -381,19 +381,53 @@ static int anetSetReuseAddr(char *err, int fd) {
...
 
-static int anetCreateSocket(char *err, int domain) {
+#define ANET_SOCKET_CLOEXEC 1
+#define ANET_SOCKET_NONBLOCK 2
+#define ANET_SOCKET_REUSEADDR 4
+static int anetCreateSocket(char *err, int domain, int type, int protocol, int flags) {
     int s;
-    if ((s = socket(domain, SOCK_STREAM, 0)) == -1) {
+
+    /* In general, SOCK_CLOEXEC won't have noticeable effect
```
Comment (emaste) [inline src/anet.c]: This comment doesn't really help me understand the context of the SOCK_CLOEXEC flag; I'm not really sure a comment is needed here at all. In both cases a flag is being set on the socket() call if possible, or being applied to the socket afterwards if not.

File: src/anet.c [hunk 2]
```diff
@@ -381,19 +381,53 @@ static int anetSetReuseAddr(char *err, int fd) {
...
+      flags &= ~ANET_SOCKET_NONBLOCK;
+    }
+#endif
+
+    if ((s = socket(domain, type, protocol)) == -1) {
         anetSetError(err, "creating socket: %s", strerror(errno));
         return ANET_ERR;
     }
 
+    if (flags & ANET_SOCKET_CLOEXEC && anetCloexec(s) == ANET_ERR) {
```
Comment (emaste) [inline src/anet.c:414]: Short-circuit `&&` doesn't seem to be existing style here, and IMO is less clear than
```
    if (flags & ANET_SOCKET_CLOEXEC) {
        if (anetCloexec(s) == ANET_ERR) {
...
```

Comment (zuiderkwast) [reply src/anet.c:414]: I think we like short-circuit style. It's in line with the style in this repo, which often prefers brevity. Double nested blocks are more complex IMO.

Comment (emaste) [reply src/anet.c:414]: OK - I was basing this comment on the (now-removed) code in this file - the `if (flags & ANET_CONNECT_NONBLOCK)` case.

File: src/anet.c (hunk 1 above)
Comment (panjf2000) [reply src/anet.c]: I'm not sure I get your point here.

Comment (zuiderkwast) [reply src/anet.c]: The comment is about why it's good to use CLOEXEC, right?

I can agree it's not the best place for this kind of advice. I suggest to put it next to or above `#define ANET_SOCKET_CLOEXEC 1`, as documentation for the flag.

File: src/anet.c [hunk 3]
```diff
@@ -381,19 +381,53 @@ static int anetSetReuseAddr(char *err, int fd) {
...
+
+    if (flags & ANET_SOCKET_NONBLOCK && anetNonBlock(err, s) == ANET_ERR) {
+        close(s);
+        return ANET_ERR;
+    }
+
     /* Make sure connection-intensive things like the benchmark tool
      * will be able to close/open sockets a zillion of times */
-    if (anetSetReuseAddr(err,s) == ANET_ERR) {
+    if (flags & ANET_SOCKET_REUSEADDR && anetSetReuseAddr(err,s) == ANET_ERR) {
```
Comment (zuiderkwast) [inline src/anet.c]: I think this comment about why we use REUSEADDR should be move to the place where it's decided that it should be used, i.e. where we set `int sockflags = ANET_SOCKET_CLOEXEC | ANET_SOCKET_REUSEADDR;` in anetTcpGenericConnect().

Btw, it's not possible to set SO_REUSEADDR when creating the socket? I guess not...

File: src/anet.c [hunk 4]
```diff
@@ -492,34 +525,6 @@ int anetTcpNonBlockBestEffortBindConnect(char *err, const char *addr, int port,
             ANET_CONNECT_NONBLOCK|ANET_CONNECT_BE_BINDING);
 }
 
-int anetUnixGenericConnect(char *err, const char *path, int flags)
-{
-    int s;
```
Comment (zuiderkwast) [inline src/anet.c:497]: I read this is unused. I'm OK with deleting it. If we ever need it again, we can find it in the git history.

**Review by zuiderkwast — COMMENTED** [review]
It looks good in general. Just a few small suggestions.

I'm not sure it has a very big impact though. It's only used for creating the listening socket and for creating outgoing connections, which is basically only for replication and cluster nodes (maybe something more?) but we don't use it for valkey-cli, valkey-benchmark and (IIRC) sentinel which use hiredis for this.

What's more important is accept that we do for each client, but we already use accept4() when it's available (setting options at the same time as accepting).

File: src/anet.c (hunk 3 above)
Comment (panjf2000) [reply src/anet.c]: > Btw, it's not possible to set SO_REUSEADDR when creating the socket? I guess not...

AFAIK, no.

File: src/anet.c (hunk 1 above)
Comment (panjf2000) [reply src/anet.c]: I'm OK with that.

**Comment by panjf2000** [comment]
> It looks good in general. Just a few small suggestions.
> 
> I'm not sure it has a very big impact though. It's only used for creating the listening socket and for creating outgoing connections, which is basically only for replication and cluster nodes (maybe something more?) but we don't use it for valkey-cli, valkey-benchmark and (IIRC) sentinel which use hiredis for this.
> 
> What's more important is accept that we do for each client, but we already use accept4() when it's available (setting options at the same time as accepting).

Speaking of which, check out #294 as well.

**Review by zuiderkwast — APPROVED** [review]

**Comment by panjf2000** [comment]
Any updates?

**Comment by zuiderkwast** [comment]
> Any updates?

I left it open for a week to let others have time to review. The `to-be-merged` label makes sure we don't forget it.

**Pull request merged**